src/lexer/mod.rs

```diff
@@ -500,6 +504,11 @@ impl<'cache, 'contents> Iterator for Lexer<'cache, 'contents> {
                 }
             },
             ('"', _) => self.lex_string(),
+            ('}', _) => {
+                self.current = '"';
+                Some((Token::InterpolateRight, self.locate()))
+            },
+            ('$', '{') => self.advance2_with(Token::InterpolateLeft),
```

Comment:
This was a rather clever implementation I didn't think of originally. It does mean that we rely on the interpolation characters not being valid in the regular expression language however. This is a relatively minor problem, though it does limit the ending character `}` from being used - we couldn't implement the `"\(foo)"` syntax with this for example.

I was wondering if this could cause worse error messages for users that try to use `${` or `}` outside of strings. The most impact it'd have I think should be mentioning an "unexpected '${' or '}' found, expected ...." which seems fine. This method doesn't work for nested interpolation like `"${foo "${bar}" baz }"` but I think that is fine.

Reply:
One approach i thought of was to keep track of the parenthesis level with the lexer, and only issue the InterpolationRight token only when the parenthesis level is zero. That would allow us to switch to a different interpolation syntax and use the token `}`outside of string interpolations

Reply:
That's a good solution - we should use it for InterpolateLeft as well

Reply:
We should be able to now still use the `}` token outside of string interpolation if we ever want to, or switch to `\()` syntax for interpolation. The lexer will issue `InterpolateRight` if the number of brackets responsible for opening the bracket to which `InterpolateRight` is mapped to that have not been closed, is the same as when `InterpolateLeft` token was last issued.

Reply:
Actually, separating open brackets into subtypes might be unnecessary. We could just keep a stack of all open brace tokens, and check if it's length matches. That might also also allow for better error messages since we could check for braces matching when popping the open braces. Not sure whether that would go beyond the scope of a lexer however

Reply:
Hmm, yeah looking over the most recent commit I do think it may be overkill to track all matching pairs. Is it really necessary to track any other tokens other than `${` and `}` for this? I was thinking the lexer could have a single `interpolate_level: u32` field or similar.